### ISSUE TITLE
[UX] Provide an option to hide the username field on registration (and generate usernames)

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#1994]
There is a setting on the Account settings page 
![loginoptions](https://cloud.githubusercontent.com/assets/1042283/16460958/1ca965f2-3df8-11e6-9cab-f86d29168519.PNG)

But if email only is selected, the user is still presented with a field to type in a username during registration. 

This issue proposes that **IF** a user is not allowed to sign in with her username, as this setting enforces, it does not make sense to still **require** a username on registration. (username could be automatically be the email itself, or generated from the email, or generated otherwise, to be decided).

And in fact it is bad for UX to do so:
* Bad UX because a user will be being asked for a field which will not be required ever again and may seem therefore a waste.
* Bad UX because a site manager who wishes to remove the need for a username field still has to resort to custom code when the above form select seems to suggest that this would be automatically not presented on registration.

Proposed issue again (updated):
**We should provide an option to hide the username field on registration, and generate usernames automatically.** This will be particularly useful if a site manager selects the "Users may log in using Email address" option, pictured above.

---
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/1763~~
PR by @jenlampton https://github.com/backdrop/backdrop/pull/1856

**Comment by bd0bd** [comment]
But if email only is selected, the user is still presented with a field to type in a username during registration.

I do not see anything confusing in this. It is correct.

**Comment by mikemccaffrey** [comment]
It is a pretty fundamental assumption in backdrop/drupal that users have usernames, so are going to need to collect them on registration even if they are not being used for log-ins.

**Comment by serundeputy** [comment]
Could/should we hide the username field with states and have it default to the email address if setting is for `Email address`?

**Comment by docwilmot** [comment]
I am thinking as the unsophisticated Backdrop newbie, who doesn't have any fundamental assumptions. 

My thinking then would be that if I configure my site to allow login with username only, I'll need a username, obviously, and an email to send my users their username. But if I configure for email only, it sort of implies that I don't need a username field. Why is it still necessary, and why is there still a field asking for it? 

Keep in mind that I am certainly not a newbie but when I selected "Email only" for the first time ever this week, I was 100% expecting the username field to disappear and was actually confused (!) when it still showed up. I assumed then that it would probably also confuse others, especially said newbies.

Unless I am way off and just easily confused, we could either use the email address as the username and simply don't print the field, or generate a username from the email, so I'm reopening this unless its conclusive that I am.

**Comment by klonos** [comment]
>My thinking then would be that if I configure my site to allow login with username only, I'll need a username, obviously, and an email to send my users their username. But if I configure for email only, it sort of implies that I don't need a username field. Why is it still necessary, and why is there still a field asking for it? 

Big 👍 there.

> Keep in mind that I am certainly not a newbie but when I selected "Email only" for the first time ever this week, I was 100% expecting the username field to disappear and was actually confused (!) when it still showed up. I assumed then that it would probably also confuse others, especially said newbies.

Same thinking here. Although not a newbie any more, I consider "my duty" to try and put myself in their shoes. That's the same reaction I'd have. I think this is a perfectly valid issue.

**Comment by klonos** [comment]
...I think that if username is a fundamentally necessary thing in order to set a user up, then hide and auto-generate if the setting is email only. @docwilmot ++

**Comment by docwilmot** [comment]
This PR hides the username field as recommended. The username is automatically set to the user's email addresss, but there are issues here:
- having a username as your email exposes your email on certain displays like in the 'author' display on nodes. Core allows emails as usernames though, so I went with this. Should core allow this? Different  issue though
- we could instead generate a unique username. Generating this from the email only partially fixes the first point above. Maybe random usernames?
- certain characters are allowed in emails but not in usernames, especially the "+" character. [This issue in Drupal](https://www.drupal.org/node/2157963) raised by none other than @jenlampton (with comments by @serundeputy too) fixed the "+" sign. We would need to fix this in Backdrop as without that fix, if a user uses an email with a "+" sign, he would get an error saying the "username" has invalid characters, even though he never put in (nor even knows what is) a username.

**Comment by jenlampton** [comment]
> We would need to fix this in Backdrop
I thought we fixed this in 1.0.0?

**Comment by docwilmot** [comment]
apparently not:
```
function user_validate_name($name) {
  ...
  if (preg_match('/[^\x{80}-\x{F7} a-z0-9@_.\'-]/i', $name)) {    <======= Note no "+"
```

**Comment by olafgrabienski** [comment]
> having a username as your email exposes your email on certain displays like in the 'author' display on nodes (...) Should core allow this?

At least, core shouldn't facilitate it, so we need a username in my opinion, but I don't like the idea random usernames. If a random name is displayed to refer to an author, it's kind of useless and even strange.

I'm sorry that I missed the start of the discussion. If I may comment it so late, I'd say that logging in and registering are just two different things. I see that the current behaviour might be confusing but then let's make the forms less confusing instead of dropping the username.

**Comment by jenlampton** [comment]
> logging in and registering are just two different things

Can you recommend a suggestion as to how we can register people without a username, without dropping the username?

What if we just use everything before the @ in the email address.

**Comment by olafgrabienski** [comment]
> What if we just use everything before the @ in the email address, instead of the whole email address?

Better than random and better than the email address. But if you know the pattern you still know the address.

> Can you recommend other suggestions as to how we can register people without a username, without dropping the username entirely?

Actually I wouldn't drop the username in the registration form as in my opinion the issue isn't really a problem: The seemingly discrepance between the setting "Users may log in using Email address" and the fact that users will be asked for it  a during registration doesn't affect prospective users (who have no particular expectations regarding the registration process) but at most the administrator user who chose the setting. Apart from that, all users will however be able to use the email address to log in, and that is what the setting says. Maybe to avoid confusion for site administrators, we could provide a help text on the settings form, like "Users will anyway be asked for a username during registration" (in better English of course).

**Comment by docwilmot** [comment]
I've created https://github.com/backdrop/backdrop-issues/issues/2486 to discuss the issue of emails as usernames.

> Better than random

Not completely true as far as displaying usernames is concerned. Usernames are often completely random. There is no word in any dictionary called "quiksketch" or "hellomrcat". So if we had used PHP to generate a random name and PHP came up with "quicksketch", there would be no difference to anyone but @quicksketch (I assume that name means something to him).

> and better than the email address. 
> But if you know the pattern you still know the address.
 
Agreed.

> doesn't affect prospective users (who have no particular expectations regarding the registration process) but at most the administrator user who chose the setting

Then why dont we ask for favorite color, or waist size too? If username is not to be used, why ask for it. Thats my issue with it.

**Comment by serundeputy** [comment]
Maybe we should just have the top two options:

<img width="403" alt="screen shot 2017-01-21 at 6 47 57 pm" src="https://cloud.githubusercontent.com/assets/6763671/22178632/5311fdf2-e00a-11e6-870e-589a8625a996.png">

and simply remove the third option `Email address` if they have to have a username then it makes sense that they can login with it.

**Comment by olafgrabienski** [comment]
> If username is not to be used, why ask for it.

Because it *is* used ... as you said, for instance in posts when display of username is enabled.

**Comment by olafgrabienski** [comment]
> simply remove the third option 'Email address'

I like that one!

**Comment by docwilmot** [comment]
> Because it is used ... as you said, for instance in posts when display of username is enabled.

😄 
touche

**Comment by docwilmot** [comment]
Then we should probably as for a "Display name" and an email address during registration, and generate the username for everyone.

**Comment by klonos** [comment]
> we could instead generate a unique username. Generating this from the email only partially fixes the first point above.

Just note that we cannot use the part of the email minus the `@domain` part because then we'd end up with duplicate usernames for `johndoe@example.com` and `johndoe@anotherexample.com` for instance.

**Comment by serundeputy** [comment]
Just get rid of the third option; the third option does more harm than good.

**Comment by jenlampton** [comment]
This was a feature we added in Backdrop because it was missing in Drupal. I'm not sure there is a good reason to remove it entirely. 

Back in the old days it was required that everyone have a username. That is not the case anymore today. Most services only require an email address and a password, and by also requiring a username in Backdrop, not only does it make the software feel dated, but it decreases the number of users that will signup. Requiring every Backdrop site to do things "the old way" would be harmful to people who only need to do things "the new way."  I don't like the idea of locking people in, especially when flexibility is supposed to be one of the reasons people would choose Backdrop in the first place.

Displaying the username on posts and comments is *optional*. For sites that don't want to use usernames, they probably also don't want to display anything here. Instead of forcing people to do things in a way that might be contrary to their vision, why not just guide them to the correct way to architect what they want.

Here are two ideas:
* If someone selects the Email only option, automatically disable the display of usernames on all node types and comments.
* If someone selects the Email only option and does choose to display usernames in nodes ro comments, add a message to the status report, notifying them that email addresses will be publically available.

**Comment by klonos** [comment]
@jenlampton perhaps what you describe could work if we are absolutely sure that there will be no other place other than node types and comments that publicly displays the username (which falls back to the email on sites that select email as the only login option if I get this right). If there is even the slightest chance of this happening, then it is a security/privacy issue.

If we decide to leave the email only login as is, then we'd need to change the label and the behavior of the "Display author and date information" checkbox under the "Display settings" vtab on content types to just "Display date information". The "Display the author picture" checkbox will need to be hidden in these cases. We'd also need to account for the "Display pictures for comment authors" checkbox in the "Comment settings" vtab.

Still though, what would a thread without any comment author names look like? I think it'd be silly.

**Comment by klonos** [comment]
...I think that requiring a username and explaining where it is being used and why makes more sense.

**Comment by docwilmot** [comment]
The question of whether emails should be used as usernames is now a different issue at https://github.com/backdrop/backdrop-issues/issues/2486. This issue is about whether the username field should still be shown when a user is registering even if "email only" is the login method. Until we decide about https://github.com/backdrop/backdrop-issues/issues/2486, if we adopt this PR, we can generate a username somehow, so lets try and separate the two issues please.

So far it seems that most of us think the username **should** remain in that situation. Is that an accurate assessment?

I've also created https://github.com/backdrop/backdrop-issues/issues/2489 to discuss changing the username field to "Display name".

**Comment by serundeputy** [comment]
Twitter, Facebook, Drupal.org, GitHub all ask for more than just your email and they all ask for some form of Name or username presumably so that they have data other than email to display with your posts.  I would guess that any service you are signing up for that lets you author information of any kind would require this.

What is the use case for email only login?  What problem does it solve?

@docwilmot what prompted you to try that setting?  What was your use case?

**Comment by docwilmot** [comment]
> What is the use case for email only login? What problem does it solve?

Its not my use case. Its in core. See first post image.

And note I have nothing against the username *per se*. Please note the issue title and first post and other posts by me. Again: **if** we offer an email-only login, it **seems** that a username shouldnt be necessary during registration. 

As per my last post, we seem to be deciding that is not true. I'm willing to let it go if there is consensus.

**Comment by serundeputy** [comment]
I understand it is in core; I'm trying to identify a use case when somebody would use it.

I think the feature we added that drupal doesn't have is that you can login with either your username or your email address.

I don't know why we added the `Email address` only version.  There must be a reason I'm trying to identify it but I can't wrap my head around it and keeping it is starting to spawn ideas that involve a lot more code.

@jenlampton do you know what the use case is for email only no username in regards to backdrop and in general to any system where the user may be publishing content?

**Comment by docwilmot** [comment]
Use case: I build a site with a registration page like this (lets call it docwilmotbook):
![fb](https://cloud.githubusercontent.com/assets/1042283/22185277/968f709a-e0b0-11e6-9d5c-72ef15be2a7a.PNG)

But I know my users will always be displayed as "Dr Firstname Surname MD". I have no need therefore to ask them for a username on registration or login.

**Comment by jenlampton** [comment]
> So far it seems that most of us think the username should remain in that situation. Is that an accurate assessment?

This will hurt adoption. Both for the sites that are using Backdrop, but also for Backdrop as a platform.  I'm not sure that's a risk we want to take when there are other viable alternatives.

**Comment by serundeputy** [comment]
Thanks @jenlampton how did you get around the username field for those sites; custom code?

**Comment by jenlampton** [comment]
>  how did you get around the username field for those sites

There was a contrib module for Drupal. For backdrop I chose the email-only option :)

**Comment by serundeputy** [comment]
but the email only option doesn't hide username from the registration form ... I thought that is what kicked this whole thing off.

**Comment by jenlampton** [comment]
Three of those sites are still in Drupal ;)

**Comment by olafgrabienski** [comment]
So use cases are the point here. The one mentioned by @docwilmot in https://github.com/backdrop/backdrop-issues/issues/1994#issuecomment-274354035 and @jenlampton's e-commerce example work without username but collect first and last name. I guess, conceptually these sites wouldn't work well without first and last name (or, more general, a display name).

Based on the assumption that we need a display name (potentially nodes and comments, system mails with tokens like "Welcome to our website, John Doe", and probably other places), we should let the admin choose which element(s) to use as a display name. Depending of the use case of a site and of the chosen Account settings this could be the usual username or a combination of other fields provided by the registration form.

I hope it would be technically possible to choose the relevant display name fields directly in the Account settings form. If an administrator then doesn't want to provide a source for a display name, the fallback should be the username, and it shouldn't be possible to offer email only registration (while email only login is still fine).

My suggestion may sound strict, but that's for a reason, at least at the moment, because so far (a) I don't see a good way to generate display names automatically, and because (b) I'm not optimistic that we find reasonable settings in various configuration places to compensate the lack of usernames.

**Meta:** The discussion of the issue has become however very complex. Does anybody dare to summarize different use cases, constraints and possible solutions?

**Comment by jenlampton** [comment]
A new summary would be good.

> conceptually these sites wouldn't work well without first and last name (or, more general, a display name).

I guess the two use cases we have are 1) sites that display some kind of name and 2) sites that don't.  Even my ecommerce site doesn't display any kind of name anywhere but on the user account page, and there the email address is suitable, as each person can only see their own page (unless they are a site admin, or a puppy trainer). The first/last information we collect on the ecommerce site is part of the reporting / credit card processing, and not needed for the website itself.

So for the second use-case, removing the username from the submitted by everywhere it appears is not only a reasonable assumption, but what people with that use-case actually want.

Why don't we just do that for sites that choose not to have a username? For sites that want a real name or other combination of fields, there are solutions for that in contrib.

**Comment by jlfranklin** [comment]
This sounds like a conversation that a site builder should have with his client, or a Web Site Steering Committee, or a development team should have among themselves.  There are trade-offs to using the email as the username, many of them cited here.

Leaving the option and documenting the ramifications of using the email as the user name is my vote.

(BTW, what happens when the user changes the email address in the profile?  Does the username follow suit?)

**Comment by docwilmot** [comment]
> The discussion of the issue has become however very complex. Does anybody dare to summarize different use cases, constraints and possible solutions?

Unfortunately, it shouldnt be complex. This issue is **not** about emails as usernames, nor display of usernames, its **just** about whether we should hide the username on the registration page if the user is not allowed to login with a username!

I've updated the first post. Please let us only decide on this issue!! 

The related issues have been separated out. See:
https://github.com/backdrop/backdrop-issues/issues/2486
https://github.com/backdrop/backdrop-issues/issues/2489
https://github.com/backdrop/backdrop-issues/issues/2494

**Comment by jenlampton** [comment]
> about whether we should hide the username on the registration page if the user is not allowed to login with a username

That seems like it should be another (separate) setting.

**Comment by olafgrabienski** [comment]
Disclaimer: this comment is not another argument but just some information which I find useful regarding how to handle a site without asking for a username.

In the project description of the Drupal module [Email Registration](https://www.drupal.org/project/email_registration) used by @jenlampton there is an interesting approach regarding the generation of usernames:

> A username is generated and assigned based on the user name part of the email address and their user ID.

On the same page, from the the tips under *Installation instructions*:

> You will probably want to change the welcome e-mail (Administer -> User Management -> User Settings) and replace instances of the token !username with !mailto

> This automatically generated username is still displayed name for posts, comments, etc. You can allow your users to change their username (...)

**Comment by docwilmot** [comment]
> That seems like it should be another (separate) setting.

Great idea! Like so?:

![login settings](https://cloud.githubusercontent.com/assets/1042283/22277025/38625cc6-e286-11e6-86d4-ea8baa7c1220.PNG)

The checkbox is hidden by `states` and only visible for "Email address" radio.
We could even further have another radio list dependent on that checkbox:
- Username or email address
- Username
- Email address
  - Hide username field on registration
    - Use email as username
    - Generate username from email
    - Generate random username

**Comment by docwilmot** [comment]
So I went ahead and did it:
https://github.com/backdrop/backdrop/pull/1763

**Comment by Dinornis** [comment]
I agree that for websites where usernames are not used or needed the registration page should not be prompting the user to create a username.  However, I am not sure if just hiding and auto-generating usernames is the right way to go about this.

If usernames are not necessary, why generating usernames at all?
If usernames are integral part of Backdrop core, why not work on removing this requirement from core?

If there is a need to auto-generate usernames, I don't think email address should be used for this, for reasons already stated by others above and as requested in the https://github.com/backdrop/backdrop-issues/issues/2486 issue.

**Comment by olafgrabienski** [comment]
Thanks for the PR! Works well, here some thoughts:

(1) We really shouldn't provide the setting "Use email as username". See also the statement on the registration form which is no longer true with this PR:

> The e-mail address is not made public and will only be used if you wish to receive a new password or wish to receive certain news or notifications by e-mail.

(2) If you change the Account settings, for instance from "Use email as username" to 
"Generate username from email", you get a mixture of 'username types'. Don't know how to handle it, but it looks kind of ugly.

(3) Usernames like "doe_1" might look ugly to the users themselves, and they might want to change it. The relevant permission which can be changed by the admin is not enabled as default for authenticated users.

(4) The email text to a new user is speaking of logging in via username even if the "Email address" option has been chosen:

> After setting your password, you will be able to log in at
> http://example.com/user in the future using:
> 
> username: doe_1
> password: Your password 

(5) With the use case "Email address" there are the above mentioned and other settings (like display of usernames as email addresses in posts or emails) without sensible defaults. It would be helpful to change them dependent on the chosen use case or to give relevant hints to the admin.

**Comment by jenlampton** [comment]
Thanks for the review @olafgrabienski !  Your points are all sensible, but I think a few of them should be separate or follow-up issues. Things like updating the email text based on the setting, is unrelated to this issue (but a really good idea!)

**Comment by olafgrabienski** [comment]
> If usernames are not necessary, why generating usernames at all?
> If usernames are integral part of Backdrop core, why not work on removing this requirement from core?

@Dinornis - I guess, usernames are necessary at the moment, and consequently we shouldn't remove them. Re 2nd question: why should we remove something that's an integral part?

**Comment by Dinornis** [comment]
My second question:
> If usernames are integral part of Backdrop core, why not work on removing this requirement from core?

was refering to @jenlampton comment:
> Back in the old days it was required that everyone have a username for everything. That is not the case anymore today. Many services only require an email address and a password, and by also requiring a username in Backdrop, not only does it make the software feel dated, but it decreases the number of users that will sign up. Requiring every Backdrop site to do things "the old way" would be harmful to people who only need to do things "the new way." I don't like the idea of locking people in, especially when flexibility is supposed to be one of the reasons people would choose Backdrop in the first place.

I understand that this is not in the scope of this PR, but it is closely linked and should be considered to come up with the right long term solutions.

**Comment by olafgrabienski** [comment]
You're welcome @jenlampton . Yes, we could separate some points, I agree, but it's also important to keep the whole picture in this quite controversial issue. So, at the moment I'm not sure which points to separate and which not. Most important in my opinion: don't introduce the possibility to use email addresses as username. Let's remove this option of the PR immediatly, we don't have to wait for a follow-up issue here.

Btw, I'm not sure if issue #2486 (don't allow users to use email address as username) has to do with the approach of the PR which allows a setting to use email addresses as username, or if it's otherly related, saying that **end users** shouldn't be allowed to do so.

**Comment by jenlampton** [comment]
> Most important in my opinion: don't introduce the possibility to use email addresses as username.

I understand that there are use-cases where people might not want this to happen, but there are also use-cases where it is perfectly fine (intranets). Or, are you just talking about doing a better job of warning people that this might be happening inadvertently? I am all in favor of that!

**Comment by Dinornis** [comment]
Yes, it may be okay to display email addresses on intranets. However they usually use both username and email fields.  Automatically populating the `Username` field with email address could cause problems down the line.

It seems like not having `Username` as a required field would solve this issue.

**Comment by quicksketch** [comment]
I took a look at the PR and I'm in agreement with others here on using email as username shouldn't be allowed as an out-of-box option.

Right now we have these 3 options:
```
+      0 => t('Use the email address as username'),
+      1 => t('Generate a username based on the email address'),
+      2 => t('Generate a random username'),
```

I'm not sure we should even bother with providing options here. I would suggest that we simply make the username be `user-[uid]` and get rid of all the code for generating random usernames. UIDs are already unique so there's no chance of a conflict. Though there will still be a bit of special code necessary because we don't get a UID until after the user row has been inserted.

OTOH this entire problem might be kicked to contrib where all the bells and whistles can be included.

**Comment by docwilmot** [comment]
> OTOH this entire problem might be kicked to contrib where all the bells and whistles can be included.

Agree with everything except this @quicksketch. As explained in the first post summary, and as experienced personally (see https://github.com/backdrop/backdrop-issues/issues/1994#issuecomment-230056830) I believe this is a UX weakness in core IMHO. If I only want users to sign in with email, and if Backdrop presents this as an option, it shouldnt still be asking for an unnecessary username. And I shouldnt have to hunt a module to hide the unnecessary field. This seems like a bug.

**Comment by jenlampton** [comment]
>  it shouldnt still be asking for an unnecessary username

In Backdrop 1.x the username is necessary whether or not people use it to log in. 

> using email as username shouldn't be allowed as an out-of-box option.

If we can't copy the email address to the username field in 1.x, this issue becomes only possible in Backdrop 2.x. (Removing milestone accordingly)

**Comment by docwilmot** [comment]
I think you missed the point.

**Comment by jenlampton** [comment]
I think you missed all the reasons this is not possible in 1.x :)

**Comment by jenlampton** [comment]
Okay, I reread all the comments and I now understand that this issue has a PR that is actually pretty darn close, so I've reinstated the 1.7 milestone. 

A few more thoughts from reading through the comments:

>  This issue is about whether the username field should still be shown when a user is registering

Facebook asks me for a name when I register. Yet I am only allowed to log-in using my email address (and I'm not allowed to log-in using my name). I must admit, also I don't really understand all the hype around this issue. It's just a setting :)

> This issue is [...] about whether we should hide the username on the registration page if the user is not allowed to login with a username

Some (I'd argue most) sites will still want to allow each user to choose the way their own author name is displayed, so this should be a setting and not an assumption.

For the PR:

* I agree with @quicksketch's recommendation of removing the option for how the username is created, that seems like something that could be added to and refined in contrib.
* I don't like the recommendation of using `user-[uid]` by default. We are trying to get away from the word "User". People dislike being called "users", the word is robotic sounding and makes the software feel kinda icky. Can we use `account-[uid]` instead?
* We'll need the label or description for that checkbox to accurately reflect that this is what's happening. Maybe something like `Automatically generate usernames (Username field will be hidden on registration form)`.
* Since this setting is related to registration (and not loging in) It should be moved to the registration fieldset. (It should also be allowed to hide/generate the username even if people are logging in via methods other than email)
* Should we add any runtime requirements checks? We could notify people via the status report if:
* * The usernames are auto-generated but people have permission to change their own usernames
* * The emails tell people to log in with un/pw when that's not allowed
* * Emails are allowed as usernames (and may be printed as author info)

**Comment by jenlampton** [comment]
I've pushed up a new PR that does the 4 things mentioned above. I also separated out the "Registration" part of the user account settings form from the "Cancellation". I'm not sure why those were grouped together, but I think this is cleaner:
<img width="1180" alt="screen shot 2017-04-30 at 9 15 27 pm" src="https://cloud.githubusercontent.com/assets/397895/25571934/5205d400-2dea-11e7-951d-67506029f8c5.png">

Please note the new setting for `Hide the username field (and automatically generate usernames).`

**Comment by docwilmot** [comment]
Not so sure about this one. I can set it so that users can *only* register with an email (by checking that new box), but on the same site I can set the "account log in" option they can only log in with a username. How would they get this username?

I think the only logical situation where username would be hidden is if you could only login with email, no (hence the dependent field in my PR)?

Confused.

**Comment by jenlampton** [comment]
> How would they get this username?

1) in the email that was sent to them when their account was created
2) from the administrator 

> I think the only logical situation where username would be hidden is if you could only login with email, no?

I have several sites where people are allowed to log in with both either username or their email address. There's a role of user (bloggers, for example) that has their accounts created by administrators, and thus does not use the user/register form. These people log in using their usernames usually. Normal folks (commenters, maybe) can create accounts using user/register, but since they don't really need usernames (and maybe you want them to appear anonymous), so do not get to specify a username.

**Comment by docwilmot** [comment]
So on registration, only email is allowed as the unique identifier, but everyone gets a username. Why would they use it, if they could login with email? If youre to mail it to them, why not let them choose it?

Sorry, its late, I'm off to bed, still confused.

**Comment by jenlampton** [comment]
> So on registration, only email is allowed as the unique identifier, but everyone gets a username.

Registration via the registration form, yes.  But there are other ways user accounts can be created. Admins can create user accounts (and still enter a username) for example.

> Why would they use it, if they could login with email?

What if they can't log in with an email? It's an admin setting, and the person who just got the email is not an admin.

> If you're going to mail it to them, why not let them choose it?

Maybe you're worried your users will enter offensive names? Maybe you want them to be anonymous?

The use-case that is the most straightforward to understand is one with different kinds of users. Some have usernames assigned to them, and want to log in using them. Some have randomly-generated usernames, and would prefer to log in using their email addresses. The setting is "Users may log in using Username or email address ". The user/login form needs to work for both these kinds of users, but the user/register form only applies to the latter.

**Comment by quicksketch** [comment]
> We are trying to get away from the word "User". People dislike being called "users", the word is robotic sounding and makes the software feel kinda icky. Can we use account-[uid] instead?

The trouble with "account" here is that you end up with a byline like "Submitted by account-1234". The account didn't submit the comment, the user did.

I can see valid uses that would hide the username field in all or only some situations:

- Usernames are never seen by end-user at all (registration, editing profile, URLs, in posts etc). Perhaps the username is replaced by separate profile fields. In this case, the field is hidden on registration and profile edit pages.
- Usernames are defaulted to a auto-value but changeable by the user later. I've seen this before on sites where users may communicate but most do not. This may be common when communication is optional but an authenticated account is required.

So basically, what I'm getting at here is we need to specifically ask how username should be handled on both registration and edit. A single option isn't sufficient.

**Comment by jenlampton** [comment]
> The trouble with "account" here is that you end up with a byline like "Submitted by account-1234". The account didn't submit the comment, the user did.

How about `member`? We can't call them "Users" to their faces. In the code is one thing...  but this is in plain sight.

> So basically, what I'm getting at here is we need to specifically ask how username should be handled on both registration and edit. A single option isn't sufficient.

We already have an option on edit. It's the permission for ` Change own username`. Is that not enough?

**Comment by jenlampton** [comment]
Naming things is hard. Naming people is harder. Here are some suggestions from around the web:
registrant, member, consumer, customer, visitor, pupil, affiliate, associate, comrade, unit, enjoyer...

what about `dragon`?

**Comment by bd0bd** [comment]
> what about dragon?

It is too original. Not clear for "normal" people :smile:

**Comment by docwilmot** [comment]
The problem with this PR is that it increases our WTF burden rather than decreasing. My original realization was simple: if a site admin sets the login settings so that users **could not** login with a username, (that is, the admin mandates an email-ONLY login), it does not make sense to still *require* a username on registration. That was the one true WTF situation we wanted to prevent here. 

We havent solved this, its still entirely possible.

But furthermore we've added another doozy: an admin can now enforce that you can **only** sign in with your username, and simultaneously prevent you from choosing one! We've not given any indication that this is likely to be a problem, and yes, you could mail it to the user, but `docwilmot` is easy to remember, `user-128732`, not.

In light of @quicksketch's frequent advice not to give people ammunition to shoot themselves in the foot, this isn't a good idea. We've given them a whole other gun.

**Comment by quicksketch** [comment]
> We already have an option on edit. It's the permission for Change own username. Is that not enough?

I hadn't thought of that. That seems like it should do it. Our existing code on the profile form:
```
    '#access' => ($register || ($user->uid == $account->uid && user_access('change own username')) || $admin),
```

Perhaps we could surface this permission on the account settings page, the same way we show permissions on the node type form.

**Comment by jenlampton** [comment]
>  if a site admin sets the login settings so that users could not login with a username, (that is, the admin mandates an email-ONLY login), it does not make sense to still require a username on registration.

@docwilmot I'm not sure why you are holding on to this idea in spite of the many use-cases we've provided where it **does** make sense to require a username. You have come up with one where it doesn't, so we are attempting to give you a feature that accounts for your use case too.

> We haven't solved this, its still entirely possible.

There's nothing to "solve" here. We don't want to prevent people from being able to require a username on registration. It's an entirely valid use case. Again: I don't want my "Name" on facebook to be my email address, or a randomly generated hash, or anything else. I want to enter it myself, while I am registering. (But I still log in using my email address).

> an admin can now enforce that you can only sign in with your username, and simultaneously prevent you from choosing one! We've not given any indication that this is likely to be a problem

It's weird, for sure, but out of the box, it isn't likely to be a problem. The email that's generated and sent automatically contains the username. One would have to change this setting to create the problem.  Slippery slope here: there are an infinite number of ways someone could configure a Backdrop site that is "likely to be a problem". We should not get into the business of telling people about all the things that might go wrong. The trick is to set it up sensibly out of the box -- which we have.

> but docwilmot is easy to remember, user-128732, not.

True, and that is a conversation that needs to be had between the person building the site and the person requesting any feature. Consider pros and cons. Make the right decision for the project and the people involved.

> Perhaps we could surface this permission on the account settings page, the same way we show permissions on the node type form.

Yep, good idea @quicksketch!

**Comment by docwilmot** [comment]
Is there any reason for wanting the UID in the generated usernames? Conversely, any security risk in exposing UIDs? I found only one post on this, https://www.drupal.org/node/949508. Couldnt we just generate completely random usernames?

**Comment by mikemccaffrey** [comment]
Hey folks, let's remember that we don't need to solve everything in core. What we want to do is to determine what the best solution is for 90% of sites, and allow the flexibility for all the other use cases to be handled in contrib modules.

Here are my suggestions:

1) Since only a small minority of sites would find it important to hide the username field, we should let them handle it in core, especially since there are so many different ways that people might want to populate the field in absence of user input. Therefore we should close this ticket.

2) It is a valid concern that users may accidentally expose their email address to the public by entering it in the username field, so we should discuss ways to mitigate the problem in #2486.

3) The email only option for logging in doesn't seem that useful. I don't see what is gained by keeping people from using a username. Maybe we should create a ticket to discuss removing it. Or maybe we should just do nothing since it is not that big of a deal.

Make sense? Can I get a second for closing this ticket? I already tried doing it once, but it didn't stick.

**Comment by jenlampton** [comment]
> Make sense? Can I get a second for closing this ticket? I already tried doing it once, but it didn't stick.

Yeah, I think that's a good idea.

> Since only a small minority of sites would find it important to hide the username field, we should let them handle it in contrib

👍 

> Maybe we should create a ticket to discuss removing this option. Or maybe we should just do nothing since it is not that big of a deal.

Since it's a big deal to some people, let's open a ticket :) I'd rather see this option removed from core (and kicked to contrib) than adding edgecase features to core to solve any UX issues it causes.
